src/Processors/Formats/Impl/Parquet/RowGroupChunkReader.cpp

```diff
@@ -0,0 +1,505 @@
+        for (size_t i = 0; i < types.size(); i++)
+        {
+            ColumnWithTypeAndName src_col = ColumnWithTypeAndName{std::move(columns.at(i)), reader_data_types.at(i), types.at(i).name};
+            // intermediate column is already casted
+            if (removeNullableOrLowCardinalityNullable(src_col.column)->getDataType()
+                == removeNullableOrLowCardinalityNullable(types.at(i).type)->getColumnType())
+            {
+                casted_columns.emplace_back(std::move(src_col.column));
+            }
+            else
+            {
+                auto casted = castColumn(src_col, types.at(i).type);
+                casted_columns.emplace_back(casted);
+            }
+        }
+        return Chunk(std::move(casted_columns), rows_read);
+    }
+    else
+        return {};
+}
+
+arrow::io::ReadRange getColumnRange(const parquet::ColumnChunkMetaData & column_metadata)
+{
+    // From ComputeColumnChunkRange() in contrib/arrow/cpp/src/parquet/file_reader.cc:
+    //  > The Parquet MR writer had a bug in 1.2.8 and below where it didn't include the
+    //  > dictionary page header size in total_compressed_size and total_uncompressed_size
+    //  > (see IMPALA-694). We add padding to compensate.
+    int64_t col_start = column_metadata.data_page_offset();
+    if (column_metadata.has_dictionary_page() && column_metadata.dictionary_page_offset() > 0
+        && col_start > column_metadata.dictionary_page_offset())
+    {
+        col_start = column_metadata.dictionary_page_offset();
+    }
+    int64_t len = column_metadata.total_compressed_size();
+    return {col_start, len};
+}
+
+
+RowGroupPrefetch::RowGroupPrefetch(
+    SeekableReadBuffer & file_, std::mutex & mutex, const parquet::ArrowReaderProperties & arrow_properties_, ThreadPool & io_pool)
+    : file(file_), file_mutex(mutex), arrow_properties(arrow_properties_)
+{
+    callback_runner = threadPoolCallbackRunnerUnsafe<ColumnChunkData>(io_pool, "ParquetRead");
+}
+void RowGroupPrefetch::prefetchRange(const arrow::io::ReadRange & range)
+{
+    if (fetched)
+        throw Exception(ErrorCodes::PARQUET_EXCEPTION, "RowGroupPrefetch: prefetchColumnChunk called after startPrefetch");
+    ranges.emplace_back(range);
+}
+void RowGroupPrefetch::startPrefetch()
+{
+    if (fetched)
+        return;
+    fetched = true;
+    auto status = arrow::io::internal::CoalesceReadRanges(
+        ranges, arrow_properties.cache_options().hole_size_limit, arrow_properties.cache_options().range_size_limit);
+    if (!status.ok())
+    {
+        throw Exception(ErrorCodes::PARQUET_EXCEPTION, "Failed to coalesce read ranges: {}", status.status().message());
+    }
+    else
+        ranges = status.MoveValueUnsafe();
+    read_range_buffers.resize(ranges.size());
+    for (size_t i = 0; i < ranges.size(); i++)
+    {
+        auto & range = ranges.at(i);
+        read_range_buffers.at(i).range = range;
+        auto task = [this, range, i]() -> ColumnChunkData
```

Comment:
Can't `this` be destroyed before `task` completes, causing use-after-free? E.g. if we finished reading row group without using all ranges for which prefetch was requested (e.g. if we skipped everything because of a condition).

Should probably `wait()` for all tasks in RowGroupPrefetch destructor. Along they way, can add something like `atomic<bool> cancelled`, check it at the start of `task`, and set it to true in destructor before waiting for the tasks.